File: codes/cpp/inheritance/fin/finalmth.cpp
#include <iostream>

class Person {
public:
    void virtual finalMethod() final {
        std::cout << "Person final method\n";
    };
    void method() {
        std::cout << "Person method\n";
    };
    void virtual virtualMethod() {
        std::cout << "Person virtual method\n";
    };
};

class Student: public Person {
public:
    //void virtual finalMethod() {};//cannot override
    void method() {
        std::cout << "Student method\n";
    };
    void virtualMethod() {
        std::cout << "Student virtual method\n";
    };
};

int main()
{
    Person* p = new Person();
    Student* s = new Student();
    Person* p2 = s;

    p->finalMethod();
    s->finalMethod();
    p2->finalMethod();

    p->method();
    s->method();
    p2->method();

    p->virtualMethod();
    s->virtualMethod();
    p2->virtualMethod();

    return 0;
}

File: codes/java/src/inheritance/fin/FinalMth.java
package inheritance.fin;

public class FinalMth {

    public static void main(String[] args) {
        Person p = new Person();
        Student s = new Student();

        p.method();
        s.method();

        p.finalMethod();
        s.finalMethod();
    }

}

class Person {
    public void method(){
        System.out.println("Person method");
    }
    public final void finalMethod(){
        System.out.println("Person final method");
    }
}

class Student extends Person {
    public void method(){
        System.out.println("Student method");
    }
    //public final void finalMethod(){}//cannot override
}
